broadlink/climate.py

```diff
@@ -238,3 +239,333 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        """
+        payload = self._send_short_payload(0)
+        if (len(payload) != 48):
+            raise ValueError(f"unexpected payload size: {len(payload)}")
+
+        # Length is 34 (0x22), the next 11 bytes are
+        # the same: bb 00 07 00 00 00 18 00 01 21 c0,
+        # bytes 0x23,0x24 are the checksum.
+        data = {}
+        data['state'] = payload[0x0d] & 0b1 == 0b1
+
+        ambient_temp = payload[0x11] & 0b00011111
+        if ambient_temp:
+            data['ambient_temp'] = (ambient_temp
+                                    + float(payload[0x21] & 0b00011111) / 10.0)
+
+        checksum = self._calculate_checksum(payload[2:0x23], 'big')
+        if (payload[0x23:0x25] != checksum):
+            logging.warning("checksum fail: calculated %s actual %s",
+                            checksum.hex(), payload[0x23:0x25].hex())
+
+        logging.debug("Received payload:\n%s", payload.hex(' '))
+
+        return data
+
+    def set_state(self, args: dict) -> bytes:
+        """Set parameters of unit.
+
+        Args:
+            args (dict): if any are missing the current value will be retrived
+                state (bool): power
+                target_temp (float): temperature set point 16<n<32
+                mode (str): cooling, heating, fan, dry, auto
+                speed (str): low, mid, high, auto
+                mute (bool):
+                turbo (bool):
+                swing_h (str): ON, OFF
+                swing_v (str): ON, OFF, 1, 2, 3, 4, 5 (fixed positions)
+                sleep (bool)
+                display (bool)
+                health (bool)
+                clean (bool)
+                mildew (bool)
+
+        Returns:
+            True for success, verified by the unit's response.
+        """
+        cmnd_0b_rmask = 0b100
+        cmnd_0c_rmask = 0b1101
+        cmnd_16 = 0b101
+
+        keys = ['state', 'mode', 'target_temp', 'speed', 'mute', 'turbo',
+                'swing_v', 'swing_h', 'sleep', 'display', 'health', 'clean',
+                'mildew']
+        unknown_keys = [key for key in args.keys() if key not in keys]
+        if len(unknown_keys) > 0:
+            raise ValueError(f"unknown argument(s) {unknown_keys}")
+
+        missing_keys = [key for key in keys if key not in args]
+        if len(missing_keys) > 0:
+            try:
+                received_state = self.get_state()
+            except RuntimeError as e:
+                if "unexpected payload size: 48" in str(e):
+                    # Occasionally a 48 byte payload gets mixed in,
+                    # a retry should suffice.
+                    received_state = self.get_state()
+                else:
+                    raise e
+            logging.debug("raw args %s", args)
+            received_state.update(args)
+            args = received_state
+            logging.debug("filled args %s", args)
+
+        PREFIX = [0xbb, 0x00, 0x06, 0x80, 0x00, 0x00, 0x0f, 0x00,
```

Comment:
I think we can assume that all requests start with `bb 00 06 80` (and responses with `bb 00 07 00`). We can abstract this with self._encode() and self._decode(), respectively.